lang/backend/src/ir2js/decls.rs

```diff
@@ -0,0 +1,235 @@
+///         case "C1":
+///             return body1;
+///         case "C2":
+///             return body2;
+///     }
+/// }
+/// ```
+impl ToJSStmt for ir::Def {
+    fn to_js_stmt(&self) -> Result<js::Stmt, BackendError> {
+        let Self { name, params, cases } = self;
+
+        // Generate self parameter
+        let mut all_params = vec![js::Param {
+            span: DUMMY_SP,
+            decorators: vec![],
+            pat: js::Pat::Ident(js::BindingIdent {
+                id: js::Ident::new("__self".into(), DUMMY_SP, SyntaxContext::empty()),
+                type_ann: None,
+            }),
+        }];
+
+        all_params.extend(params.iter().map(|p| js::Param {
+            span: DUMMY_SP,
+            decorators: vec![],
+            pat: js::Pat::Ident(js::BindingIdent {
+                id: js::Ident::new(p.clone().into(), DUMMY_SP, SyntaxContext::empty()),
+                type_ann: None,
+            }),
+        }));
+
+        // Generate switch statement on self.tag
+        let cases = cases
+            .iter()
+            .map(|case| case.to_js_switch_case("__self"))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let body_stmts = vec![js::Stmt::Switch(js::SwitchStmt {
+            span: DUMMY_SP,
+            discriminant: Box::new(js::Expr::Member(js::MemberExpr {
+                span: DUMMY_SP,
+                obj: Box::new(js::Expr::Ident(js::Ident::new(
+                    "__self".into(),
+                    DUMMY_SP,
+                    SyntaxContext::empty(),
+                ))),
+                prop: js::MemberProp::Ident(js::IdentName { span: DUMMY_SP, sym: "tag".into() }),
+            })),
+            cases,
+        })];
+
+        Ok(js::Stmt::Decl(js::Decl::Fn(js::FnDecl {
+            ident: js::Ident::new(name.clone().into(), DUMMY_SP, SyntaxContext::empty()),
+            declare: false,
+            function: Box::new(js::Function {
+                params: all_params,
+                decorators: vec![],
+                span: DUMMY_SP,
+                ctxt: SyntaxContext::empty(),
+                body: Some(js::BlockStmt {
+                    span: DUMMY_SP,
+                    ctxt: SyntaxContext::empty(),
+                    stmts: body_stmts,
+                }),
+                is_generator: false,
+                is_async: false,
+                type_params: None,
+                return_type: None,
+            }),
+        })))
+    }
+}
+
+/// Input:
+///
+/// ```text
+/// codef name(params) {
+///     .d1(args) => body1,
+///     .d2(args) => body2,
+/// }
+/// ```
+///
+/// Output:
+///
+/// ```js
+/// function name(params) {
+///     return {
+///         d1: function(args) { return body1; },
+///         d2: function(args) { return body2; }
+///     };
+/// }
+/// ```
+impl ToJSStmt for ir::Codef {
+    fn to_js_stmt(&self) -> Result<js::Stmt, BackendError> {
+        let Self { name, params, cases } = self;
+        let js_params: Vec<_> = params
```

Comment:
The code to compute params is duplicated among several functions in this module. Should be unified.